In [ ]:
import numpy as np
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, LSTM, Bidirectional, Dense

num_words = 10000
maxlen = 200

(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=num_words)
X_train = pad_sequences(X_train, maxlen=maxlen)
X_test = pad_sequences(X_test, maxlen=maxlen)

17464789/17464789 [==============================] - 0s 0us/step


# RNN


In [ ]:
model_rnn = Sequential()
model_rnn.add(Embedding(num_words, 32))
model_rnn.add(SimpleRNN(64))
model_rnn.add(Dense(1, activation='sigmoid'))

model_rnn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model_rnn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, None, 32)          320000    
                                                                 
 simple_rnn (SimpleRNN)      (None, 64)                6208      
                                                                 
 dense (Dense)               (None, 1)                 65        
                                                                 
Total params: 326,273
Trainable params: 326,273
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model_rnn.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=128)

loss, accuracy = model_rnn.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: {:.2f}%'.format(accuracy * 100))

Epoch 1/10
196/196 [==============================] - 31s 148ms/step - loss: 0.6542 - accuracy: 0.5986 - val_loss: 0.4651 - val_accuracy: 0.7905
Epoch 2/10
196/196 [==============================] - 24s 125ms/step - loss: 0.4431 - accuracy: 0.8042 - val_loss: 0.4023 - val_accuracy: 0.8294
Epoch 3/10
196/196 [==============================] - 24s 124ms/step - loss: 0.3399 - accuracy: 0.8556 - val_loss: 0.4183 - val_accuracy: 0.8148
Epoch 4/10
196/196 [==============================] - 25s 129ms/step - loss: 0.2982 - accuracy: 0.8798 - val_loss: 0.4049 - val_accuracy: 0.8368
Epoch 5/10
196/196 [==============================] - 24s 122ms/step - loss: 0.2440 - accuracy: 0.9065 - val_loss: 0.4225 - val_accuracy: 0.8212
Epoch 6/10
196/196 [==============================] - 25s 129ms/step - loss: 0.2140 - accuracy: 0.9193 - val_loss: 0.4264 - val_accuracy: 0.8293
Epoch 7/10
196/196 [==============================] - 28s 145ms/step - loss: 0.1582 - accuracy: 0.9446 - val_loss: 0.4507 - val_ac

**RNN**

**5 епох**, виконання 2 хвилини, точність 83.04%

**10 епох**, виконання 4 хвилини, точність 77.06%.
Під час тренування спостерігався overfitting. Пік точності було досягнуто на 7 епосі - 84%

# LSTM

In [ ]:
model_lstm = Sequential()
model_lstm.add(Embedding(num_words, 32))
model_lstm.add(LSTM(64))
model_lstm.add(Dense(1, activation='sigmoid'))

model_lstm.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model_lstm.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_1 (Embedding)     (None, None, 32)          320000    
                                                                 
 lstm (LSTM)                 (None, 64)                24832     
                                                                 
 dense_1 (Dense)             (None, 1)                 65        
                                                                 
Total params: 344,897
Trainable params: 344,897
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model_lstm.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=128)

loss, accuracy = model_lstm.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: {:.2f}%'.format(accuracy * 100))


Epoch 1/10
196/196 [==============================] - 89s 439ms/step - loss: 0.4721 - accuracy: 0.7738 - val_loss: 0.3139 - val_accuracy: 0.8672
Epoch 2/10
196/196 [==============================] - 104s 533ms/step - loss: 0.2433 - accuracy: 0.9054 - val_loss: 0.3216 - val_accuracy: 0.8720
Epoch 3/10
196/196 [==============================] - 103s 526ms/step - loss: 0.1862 - accuracy: 0.9328 - val_loss: 0.3160 - val_accuracy: 0.8706
Epoch 4/10
196/196 [==============================] - 104s 529ms/step - loss: 0.1531 - accuracy: 0.9458 - val_loss: 0.4486 - val_accuracy: 0.8527
Epoch 5/10
196/196 [==============================] - 103s 525ms/step - loss: 0.1302 - accuracy: 0.9552 - val_loss: 0.3885 - val_accuracy: 0.8629
Epoch 6/10
196/196 [==============================] - 101s 513ms/step - loss: 0.0903 - accuracy: 0.9716 - val_loss: 0.6018 - val_accuracy: 0.8466
Epoch 7/10
196/196 [==============================] - 82s 418ms/step - loss: 0.0751 - accuracy: 0.9758 - val_loss: 0.4826 - v

**LSTM**

**5 епох**, 7 хвилин, точність 85.98%

**10 епох**, 17 хвилин, точність 84.84%.
Спостерігається незначний спад точності після початкового підвищення

# Двостороння мережа (Bidirectional)

In [ ]:
model_bidirectional = Sequential()
model_bidirectional.add(Embedding(num_words, 32))
model_bidirectional.add(Bidirectional(LSTM(64)))
model_bidirectional.add(Dense(1, activation='sigmoid'))

model_bidirectional.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model_bidirectional.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, None, 32)          320000    
                                                                 
 bidirectional (Bidirectiona  (None, 128)              49664     
 l)                                                              
                                                                 
 dense_2 (Dense)             (None, 1)                 129       
                                                                 
Total params: 369,793
Trainable params: 369,793
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model_bidirectional.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=128)

loss, accuracy = model_bidirectional.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: {:.2f}%'.format(accuracy * 100))

Epoch 1/10
196/196 [==============================] - 149s 740ms/step - loss: 0.4703 - accuracy: 0.7699 - val_loss: 0.3274 - val_accuracy: 0.8655
Epoch 2/10
196/196 [==============================] - 145s 737ms/step - loss: 0.2471 - accuracy: 0.9058 - val_loss: 0.3040 - val_accuracy: 0.8704
Epoch 3/10
196/196 [==============================] - 146s 748ms/step - loss: 0.1847 - accuracy: 0.9316 - val_loss: 0.3222 - val_accuracy: 0.8692
Epoch 4/10
196/196 [==============================] - 137s 698ms/step - loss: 0.1518 - accuracy: 0.9475 - val_loss: 0.3467 - val_accuracy: 0.8602
Epoch 5/10
196/196 [==============================] - 147s 750ms/step - loss: 0.1267 - accuracy: 0.9555 - val_loss: 0.4532 - val_accuracy: 0.8486
Epoch 6/10
196/196 [==============================] - 145s 739ms/step - loss: 0.1015 - accuracy: 0.9654 - val_loss: 0.3887 - val_accuracy: 0.8576
Epoch 7/10
196/196 [==============================] - 145s 742ms/step - loss: 0.0846 - accuracy: 0.9718 - val_loss: 0.4458 -

**Bidirectional**

**5 епох**, 12 хвилин, точність 85.76%

**10 епох**, виконання 25 хвилин, точність 84.94%. Попри зниження точності після перших кількох епох, модель Bidirectional все ще показує достатньо високу точність

# Глибока мережа (Deep)


In [ ]:
model_deep = Sequential()
model_deep.add(Embedding(num_words, 32))
model_deep.add(LSTM(64, return_sequences=True))
model_deep.add(LSTM(64))
model_deep.add(Dense(1, activation='sigmoid'))

model_deep.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model_deep.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_3 (Embedding)     (None, None, 32)          320000    
                                                                 
 lstm_2 (LSTM)               (None, None, 64)          24832     
                                                                 
 lstm_3 (LSTM)               (None, 64)                33024     
                                                                 
 dense_3 (Dense)             (None, 1)                 65        
                                                                 
Total params: 377,921
Trainable params: 377,921
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model_deep.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=128)

loss, accuracy = model_deep.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: {:.2f}%'.format(accuracy * 100))

Epoch 1/10
196/196 [==============================] - 223s 1s/step - loss: 0.4560 - accuracy: 0.7718 - val_loss: 0.3188 - val_accuracy: 0.8656
Epoch 2/10
196/196 [==============================] - 214s 1s/step - loss: 0.2531 - accuracy: 0.9003 - val_loss: 0.3200 - val_accuracy: 0.8669
Epoch 3/10
196/196 [==============================] - 215s 1s/step - loss: 0.1890 - accuracy: 0.9301 - val_loss: 0.3745 - val_accuracy: 0.8650
Epoch 4/10
196/196 [==============================] - 212s 1s/step - loss: 0.1521 - accuracy: 0.9460 - val_loss: 0.4169 - val_accuracy: 0.8536
Epoch 5/10
196/196 [==============================] - 213s 1s/step - loss: 0.1202 - accuracy: 0.9598 - val_loss: 0.4419 - val_accuracy: 0.8449
Epoch 6/10
196/196 [==============================] - 216s 1s/step - loss: 0.1036 - accuracy: 0.9640 - val_loss: 0.4043 - val_accuracy: 0.8522
Epoch 7/10
196/196 [==============================] - 214s 1s/step - loss: 0.0703 - accuracy: 0.9780 - val_loss: 0.5481 - val_accuracy: 0.8524

**Deep**

**5 епох**, 19 хвилин, точність 85.55%

**10 епох**, виконання 38 хвилин, точність 84.61%.
 Після певного зростання точності протягом перших епох, точність на валідаційному наборі стає стійкою та показує добрі результати

# Висновки

**Після тренування моделей протягом 5 епох**


**LSTM** та двостороння мережа **Bidirectional** показують кращі результати, оскільки вони можуть краще усвідомлювати довгострокові залежності в тексті. Однак, різниця у точності між цими моделями є несуттєвою.

**Загальні**

**RNN**: При збільшенні кількості епох з 5 до 10 спостерігається overfitting моделі, що проявляється у зниженні точності на валідаційному наборі даних. При 7 епохах досягається пік точності, але загальна точність (77.06%) є нижчою в порівнянні з іншими моделями.

**LSTM**: При 5 епохах модель досягає високої точності (85.98%), але при подовженні тренування до 10 епох точність незначно знижується (84.84%). Це може свідчити про наявність незначного overfitting після певної кількості епох.

**Bidirectional**: Модель показує добрі результати при обох кількостях епох (85.76% при 5 епохах і 84.94% при 10 епохах). Хоча точність знижується після перших кількох епох, вона все ще залишається на високому рівні, що свідчить про ефективність моделі.

**Deep**: Точність моделі Deep також залишається на високому рівні незалежно від кількості епох (85.55% при 5 епохах і 84.61% при 10 епохах). Після певного зростання точності протягом перших епох, модель стає стійкою і показує добрі результати.


Загалом, можна зробити висновок, що моделі **LSTM**, **Bidirectional** і **Deep** показують добрі результати у класифікації рецензій, з високою точністю навіть після певного зниження в деяких випадках. Проте, **Bidirectional** і особливо **Deep** вимагають значно більше часу для навчання